### PR TITLE
Feature/huge refactor

### PR DESCRIPTION
**Pull request by arekpaterak**

### CODE REVIEW

File: src/ml_for_science_articles/data/utils.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ def split_by_year(
     train_end_year: int,
     val_end_year: int,
     date_column: str = "publication_date",
-) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    with_test: bool = False,
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame | None]:
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/data/utils.py]: Maybe `Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame] | Tuple[pd.DataFrame, pd.DataFrame]` ? this way we will not have to make _ at assigning values

File: notebooks/3-embed-with-pretrained-model.ipynb
Comment (Tosiekdev) [inline notebooks/3-embed-with-pretrained-model.ipynb:1]: `train_df, val_df, _ = split_by_year(df, train_end_year=2020, val_end_year=2021)` look at the comment for split_by_year in data/utils

File: notebooks/4-evaluate.ipynb
Comment (Tosiekdev) [inline notebooks/4-evaluate.ipynb:1]: Consider using train_df.apply(Paper.from_df_row, axis=1).values instead of list comprahension it is slightly faster

File: src/ml_for_science_articles/evaluator.py [hunk 2]
```diff
@@ -7,100 +7,120 @@
...
 
 class RecommendationsEvaluator:
     """
     Evaluates recommender system performance using various metrics.
     """
 
-    def __init__(self, df: pd.DataFrame, paper_id_col: str = "paperId"):
-        """
-        Initialize evaluator.
+    def __init__(self):
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/evaluator.py]: if pass then unnecessary

File: src/ml_for_science_articles/evaluator.py [hunk 3]
```diff
@@ -7,100 +7,120 @@
...
+            Dictionary of evaluation metrics
+        """
+        return {
+            "references_overlap_score": self.references_overlap_score(
+                relevant_papers_list, recommendations_list
+            ),
+            "mean_reciprocal_rank": self.mean_reciprocal_rank(
+                relevant_papers_list, recommendations_list
+            ),
+            "precision_at_10": self.precision_at_k(
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/evaluator.py]: we may want to parametrize this

File: src/ml_for_science_articles/evaluator.py [hunk 4]
```diff
@@ -7,100 +7,120 @@
...
+            Mean references overlap score (0 to 1)
         """
-        if not recommendations:
-            return 0.0
+        scores = []
 
-        relevant_papers = self._get_paper_references(query_paper_id)
+        for relevant_papers, recommendations in zip(relevant_papers_list, recommendations_list):
+            if not recommendations:
+                scores.append(0.0)
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/evaluator.py]: what if there is no relevant papers? i think in this case we should append NaN

File: src/ml_for_science_articles/evaluator.py [hunk 5]
```diff
@@ -7,100 +7,120 @@
...
-        if not relevant_papers:
+        rr_scores = []
+        for relevant_papers, recommendations in zip(relevant_papers_list, recommendations_list):
+            rr = self._reciprocal_rank(relevant_papers, recommendations)
+            if not np.isnan(rr):
+                rr_scores.append(rr)
+        if not rr_scores:
             return np.nan
-
+        return np.nanmean(rr_scores)
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/evaluator.py]: we cannot have nans in rr_scores so np.mean is enough

File: src/ml_for_science_articles/evaluator.py [hunk 6]
```diff
@@ -7,100 +7,120 @@
...
             k: Number of top recommendations to consider
 
         Returns:
             Precision@k score (0 to 1)
         """
-        recommendations_k = recommendations[:k]
-        return self.references_overlap_score(query_paper_id, recommendations_k)
+        recommendations_k_list = [recommendations[:k] for recommendations in recommendations_list]
+        return self.references_overlap_score(relevant_papers_list, recommendations_k_list)
+
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/evaluator.py]: i suppose github needs one more new line to be beautiful

File: src/ml_for_science_articles/models/__init__.py [hunk 7]
```diff
@@ -6,12 +6,12 @@
...
-from ml_for_science_articles.models.types import Paper
+from ml_for_science_articles.types import TitleAndAbstract
 
 __all__ = [
     "BaseEmbedder",
     "SentenceTransformerEmbedder",
     "Specter2Embedder",
     "create_embedder",
-    "Paper",
+    "TitleAndAbstract",
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/models/__init__.py]: Why no Paper?

File: src/ml_for_science_articles/models/specter2_embedder.py [hunk 8]
```diff
@@ -98,8 +103,8 @@ def encode(
         with torch.no_grad():
             for i in iterator:
                 batch_start = i * batch_size
-                batch_end = min((i + 1) * batch_size, len(text_batch))
-                batch_texts = text_batch[batch_start:batch_end]
+                batch_end = min((i + 1) * batch_size, len(texts))
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/models/specter2_embedder.py]: slicing is safe, no need to calculate batch end

File: src/ml_for_science_articles/preprocessors.py [hunk 9]
```diff
@@ -6,32 +6,50 @@
...
-class TextPreprocessor:
+
+class DataFramePreprocessor:
     """
-    Handles text cleaning and preprocessing for document corpus.
+    Preprocesses dataframes containing scientific articles.
     """
 
-    def __init__(
+    def __init__(self):
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/preprocessors.py]: redundant

File: src/ml_for_science_articles/preprocessors.py [hunk 10]
```diff
@@ -6,32 +6,50 @@
...
-        self.remove_numbers = remove_numbers
-        self.title_weight = title_weight
+        assert (
+            title_col in df.columns
+        ), f"Title column '{title_col}' not found in dataframe."
+        assert (
+            abstract_col in df.columns
+        ), f"Abstract column '{abstract_col}' not found in dataframe."
+
+        assert df[title_col].notna().all(), "Some titles are missing."
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/preprocessors.py]: we are doing preprocessing, can't we do dropna()?

File: src/ml_for_science_articles/preprocessors.py [hunk 11]
```diff
@@ -62,18 +82,60 @@ def _flatten_references_column(self, df: pd.DataFrame, references_col: str) -> p
...
         """
         # from list of {'paperId': ...} to list of paperId strings
-        
+
         def extract_references(refs):
             if refs is None:
                 return []
             if not isinstance(refs, list):
                 return []
             if len(refs) == 0:
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/preprocessors.py]: imo redundant

File: src/ml_for_science_articles/recommender.py [hunk 12]
```diff
@@ -2,96 +2,116 @@
...
+from ml_for_science_articles.vectorizers import BaseVectorizer
 
 
 class VectorRecommender:
-    """
-    Recommender using text vectorization.
-
-    Can plug in any vectorizer that implements BaseVectorizer interface.
-    """
-
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/recommender.py:21]: why we got rid of documentation? it applies to the whole file

File: src/ml_for_science_articles/recommender.py [hunk 13]
```diff
@@ -2,96 +2,116 @@
...
-        self.text_col = text_col
-        self.paper_id_col = paper_id_col
-        self.title_col = title_col
-        self.abstract_col = abstract_col
-
-        self.vectors: Optional[np.ndarray] = None
-        self.is_fitted = False
+        self.papers = papers
+        self.vectors = vectors
+        self.similarity_fn_name = similarity_fn_name
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/recommender.py]: unused variable

File: src/ml_for_science_articles/recommender.py [hunk 14]
```diff
@@ -2,96 +2,116 @@
...
-        if len(matches) == 0:
-            raise ValueError(f"Paper ID '{paper_id}' not found in corpus.")
-        return matches[0]
+        n_queries = query_vectors.shape[0]
+
+        assert (
+            len(paper_ids) == n_queries
+        ), "Number of paper IDs must match number of query vectors."
+
+        similarities = cosine_similarity(query_vectors, self.vectors)
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/recommender.py:70]: maybe here use used function?

File: src/ml_for_science_articles/recommender.py [hunk 15]
```diff
@@ -2,96 +2,116 @@
...
+
+            # Sort the top-k indices by similarity score
+            row_indices = np.arange(n_queries)[:, np.newaxis]
+            sorted_positions = np.argsort(
+                similarities[row_indices, top_indices], axis=1
+            )[:, ::-1]
+            top_indices = np.take_along_axis(top_indices, sorted_positions, axis=1)
+        else:
+            # Full sort if requesting more than available
+            top_indices = np.argsort(similarities, axis=1)[:, ::-1][:, :top_k]
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/recommender.py]: aren't we then just taking ALL indicies? so [:, :top_k] is redundant?

File: src/ml_for_science_articles/recommender.py [hunk 16]
```diff
@@ -104,103 +124,136 @@ def recommend_by_paper_id(
...
+        # Exclude the specified index if provided
+        if exclude_idx is not None:
+            similarities[exclude_idx] = -np.inf
+
+        # Use argpartition for O(n) instead of O(n log n) when top_n << n
+        if top_k < len(similarities):
+            top_indices = np.argpartition(similarities, -top_k)[-top_k:]
+            top_indices = top_indices[np.argsort(similarities[top_indices])[::-1]]
+        else:
+            top_indices = similarities.argsort()[::-1][:top_k]
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/recommender.py]: once again, we are just getting all indices i guess

File: src/ml_for_science_articles/vectorizers/count_vectorizer.py [hunk 17]
```diff
@@ -14,129 +15,132 @@
...
-        max_df: float = 0.8,
+        min_df_absolute: int = 1,
+        max_df_fraction: float = 1.0,
         ngram_range: tuple = (1, 1),
-        stop_words: str = 'english',
+        stop_words: str = "english",
         binary: bool = False,
-        normalize_by: str | None = None,
-        **kwargs
+        normalize_by: Literal["l1", "l2"] | None = None,
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/vectorizers/count_vectorizer.py]: max is also valid normalize

File: src/ml_for_science_articles/vectorizers/tfidf_vectorizer.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+    def __init__(
+        self,
+        max_features: int = 5000,
+        min_df_absolute: int = 1,
+        max_df_fraction: float = 1.0,
+        ngram_range: tuple = (1, 1),
+        stop_words: str = "english",
+        use_idf: bool = True,
+        normalize_by: Literal["l1", "l2"] | None = "l2",
```
Comment (Tosiekdev) [inline src/ml_for_science_articles/vectorizers/tfidf_vectorizer.py:31]: once again, max

File: tests/test_evaluator.py [hunk 19]
```diff
@@ -277,40 +279,68 @@ def test_no_relevant_recommendations(self, evaluator):
         Query refs: ref_A, ref_B, ref_C
         Recommendations: [other_X, other_Y]
         No matches found
-        Expected: 0.0
+        Expected: NaN (no RR for any query)
```
Comment (Tosiekdev) [inline tests/test_evaluator.py]: i know this passes, but maybe we have to change implementation of MRR? no relevant recomendation i think should result in the score 0

File: tests/test_evaluator.py [hunk 20]
```diff
@@ -277,40 +279,68 @@ def test_no_relevant_recommendations(self, evaluator):
...
+                Recommendation(paper_id="other_X", title="Other X"),
+                Recommendation(paper_id="other_Y", title="Other Y"),
+            ]
         ]
-        mrr = evaluator.mean_reciprocal_rank("query", recommendations)
-        assert mrr == 0.0
+        mrr = evaluator.mean_reciprocal_rank(relevant_papers_list, recommendations_list)
+        assert np.isnan(mrr)
 
     def test_empty_recommendations(self, evaluator):
```
Comment (Tosiekdev) [inline tests/test_evaluator.py]: similarly empty recommendations

File: tests/vectorizers/test_count_vectorizer.py [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+        X = vectorizer.fit_transform(sample_texts)
+
+        # Each row should have L2 norm of 1
+        X_dense = X.toarray()
+        norms = np.linalg.norm(X_dense, axis=1)
+        np.testing.assert_array_almost_equal(
+            norms, np.ones(len(sample_texts)), decimal=5
+        )
+
+    def test_normalization_l1(self, sample_texts):
```
Comment (Tosiekdev) [inline tests/vectorizers/test_count_vectorizer.py:131]: if you decide to add max metric, please remember to test it

File: src/ml_for_science_articles/models/__init__.py (hunk 7 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/models/__init__.py]: Why TitleAndAbstract? Both Paper and TitleAndAbstract are outside of models, my fault

File: src/ml_for_science_articles/evaluator.py (hunk 5 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/evaluator.py]: I changed reciprocal_rank, it returns NaN when there are no relevant_papers

File: src/ml_for_science_articles/recommender.py (hunk 12 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/recommender.py:21]: `SemanticSearchRecommender` works differently than VectorRecommender. I have some alternative design on local branch, we will make both of them consistent in the next step after this PR

File: src/ml_for_science_articles/recommender.py (hunk 14 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/recommender.py:70]: It needs changes anyway because it is that computation that results in 32 GB memory usage for queries from training set. Maybe the vector store will handle this functionality

File: src/ml_for_science_articles/recommender.py (hunk 15 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/recommender.py]: Nice catch

File: src/ml_for_science_articles/vectorizers/tfidf_vectorizer.py (hunk 18 above)
Comment (arekpaterak) [reply src/ml_for_science_articles/vectorizers/tfidf_vectorizer.py:31]: Max is not handled by TfidfVectorizer, [sklearn docs](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html).

image

File: tests/vectorizers/test_count_vectorizer.py (hunk 21 above)
Comment (arekpaterak) [reply tests/vectorizers/test_count_vectorizer.py:131]: I added a new test

File: tests/test_evaluator.py (hunk 19 above)
Comment (arekpaterak) [reply tests/test_evaluator.py]: You are right

**Review by Tosiekdev — APPROVED** [review]
